### DISCUSSION TITLE
Reduce cost of coupled model (focus on waves)

### DISCUSSION THREAD

**Discussion by pjpegion** [ufs-community/ufs-weather-model#885] [Enhancements]
I am trying to figure out different ways to speed up the coupled model, one thing I am working on is moving from a 75 layer ocean model to a 41 layer model, which reduces the cost of the ocean by about 45%.  But only about 7% for the full coupled model.  One thing that I noticed is that the wave model now costs more than the ocean model!  Looking at the job sequence, I see that the wave model is running at the short-time step with the atmosphere and ice-models.  
  Would it be possible for me to move the wave model to run at the ocean time-step to reduce its cost?  I don't know enough about how the nems.configure works to do this myself and could use help suggestions on how to configure this, and I assume something would need to change in the wave model namelist too.
  Thoughts?

**Comment by junwang-noaa** [comment]
@pjpegion what resolutions are you working on? @aliabdolali Ali is working on creating low resolution ww3 grid, what will reduce the computation and total number tasks for ww3. @aliabdolali do you have any updates on this? Thanks

**Reply by aliabdolali** [reply]
You can increase the wave model speed substantially by decreasing the spectral resolution.
In the ww3_grid.inp, you can change the number of spectral bins and number of directions. For example, in the existing glo_1deg, you can change it from   `1.07  0.035  50  36  0.5` to   `1.07  0.035  25  24  0.5`.
Another way would be to make the spatial grid resolution higher. Then you can justify the time steps and make them coarser (here is the link to [HOW TO SET TIME STEPS](https://github.com/NOAA-EMC/WW3/wiki/FAQs-page#CFL-condition)).

**Reply by pjpegion** [reply]
I'm looking at the section on how to set time-steps, and how do I know what the maximum latitude is?  If I assume it is 90-degrees, then the maximum time-step is zero.  I tried looking the mod_def.gwes_30m file, but I have no idea what the format of that file is, and cannot find that information in the documentation.

**Reply by aliabdolali** [reply]
> I'm looking at the section on how to set time-steps, and how do I know what the maximum latitude is? If I assume it is 90-degrees, then the maximum time-step is zero. I tried looking the mod_def.gwes_30m file, but I have no idea what the format of that file is, and cannot find that information in the documentation.

The max latitude is 80 for gwes_30m and the time steps are   1800.  450.  900.  10.

**Reply by pjpegion** [reply]
Thanks!  But looking at this run sequence.  it looks like the wave model's time-step is 300 seconds, am I missing something?

# CMEPS warm run sequence
runSeq::
@1800
   MED med_phases_prep_ocn_avg
   MED -> OCN :remapMethod=redist
   OCN -> WAV
   WAV -> OCN :srcMaskValues=1
   OCN
   @300
     MED med_phases_prep_atm
     MED med_phases_prep_ice
     MED -> ATM :remapMethod=redist
     MED -> ICE :remapMethod=redist
     WAV -> ATM :srcMaskValues=1
     ATM -> WAV
     ICE -> WAV
     ATM
     ICE
     WAV
     ATM -> MED :remapMethod=redist
     MED med_phases_post_atm
     ICE -> MED :remapMethod=redist
     MED med_phases_post_ice
     MED med_phases_prep_ocn_accum
   @
   OCN -> MED :remapMethod=redist
   MED med_phases_post_ocn
   MED med_phases_restart_write
@

**Reply by aliabdolali** [reply]
if the coupling time step is smaller than model time steps, it rules out the model time step.
Is this a regtest or a real application? I mean it does not make sense to use such coarse wave grids and keep coupling time step small.

**Comment by pjpegion** [comment]
I'm running C384/0.25 degree, but even with a lower resolution version,  is it necessary to have it run at the fast, atmosphere time-step. Wind waves travel ~10m/s so the longer ocean time-step should be fine.

**Reply by JessicaMeixner-NOAA** [reply]
@SMoorthi-emc I believe there's a reason for the atm model to have waves running at the atm time step, correct?

**Comment by SMoorthi-emc** [comment]
Yes, because when two-way coupled, the wave model gives surface roughness
over open ocean used in physics.
Moorthi

On Fri, Oct 22, 2021 at 12:09 PM Jessica Meixner ***@***.***>
wrote:

> @SMoorthi-emc <https://github.com/SMoorthi-emc> I believe there's a
> reason for the atm model to have waves running at the atm time step,
> correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/discussions/885#discussioncomment-1522185>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALLVRYRDM23A54HMEFVKHZ3UIGEBBANCNFSM5GQVNIKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Dr. Shrinivas Moorthi
Research Meteorologist
Modeling and Data Assimilation Branch
Environmental Modeling Center / National Centers for Environmental
Prediction
5830 University Research Court - (W/NP23), College Park MD 20740 USA
Tel: (301)683-3718

e-mail: ***@***.***
Phone:  (301) 683-3718   Fax: (301) 683-3718

**Reply by pjpegion** [reply]
@SMoorthi-emc but does the roughness length really need to be updated every physics time-step, and do the wave need to have winds updated so frequently?   Running at the ocean time-step is a cost savings of 6x.

**Comment by yangfanglin** [comment]
I recall  different spectral bins and numbers of directions were tested
(reduced) to speed up WAVE for GFS.v16 implementation.  I was not they
should not be reduced any further at that time to maintain science
integrity.

Fanglin

On Fri, Oct 22, 2021 at 11:26 AM Ali.Abdolali ***@***.***>
wrote:

> You can increase the wave model speed substantially by decreasing the
> spectral resolution.
> In the ww3_grid.inp, you can change the number of spectral bins and number
> of directions. For example, in the existing glo_1deg, you can change it
> from 1.07 0.035 50 36 0.5 to 1.07 0.035 25 24 0.5.
> Another way would be to make the spatial grid resolution higher. Then you
> can justify the time steps and make them coarser (here is the link to HOW
> TO SET TIME STEPS
> <https://github.com/NOAA-EMC/WW3/wiki/FAQs-page#CFL-condition>).
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/discussions/885#discussioncomment-1521983>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AKY5N2PUUBO2KYY4UT4P7GTUIF7C7ANCNFSM5GQVNIKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
*Fanglin Yang, Ph.D.*
*Chief, Model Physics Group*
*Modeling and Data Assimilation Branch*

*NOAA/NWS/NCEP Environmental Modeling Center*

*https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/
<https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/>*

**Reply by aliabdolali** [reply]
> I recall different spectral bins and numbers of directions were tested (reduced) to speed up WAVE for GFS.v16 implementation. I was not they should not be reduced any further at that time to maintain science integrity. Fanglin
> […](#)
> On Fri, Oct 22, 2021 at 11:26 AM Ali.Abdolali ***@***.***> wrote: You can increase the wave model speed substantially by decreasing the spectral resolution. In the ww3_grid.inp, you can change the number of spectral bins and number of directions. For example, in the existing glo_1deg, you can change it from 1.07 0.035 50 36 0.5 to 1.07 0.035 25 24 0.5. Another way would be to make the spatial grid resolution higher. Then you can justify the time steps and make them coarser (here is the link to HOW TO SET TIME STEPS <https://github.com/NOAA-EMC/WW3/wiki/FAQs-page#CFL-condition>). — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#885 (reply in thread)](https://github.com/ufs-community/ufs-weather-model/discussions/885#discussioncomment-1521983)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AKY5N2PUUBO2KYY4UT4P7GTUIF7C7ANCNFSM5GQVNIKA> . Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
> -- *Fanglin Yang, Ph.D.* *Chief, Model Physics Group* *Modeling and Data Assimilation Branch* *NOAA/NWS/NCEP Environmental Modeling Center* *https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/ <https://www.emc.ncep.noaa.gov/gmb/wx24fy/fyang/>*

Right, it is ok for regtesting, otherwise, they should be higher in operational setups.

**Comment by SMoorthi-emc** [comment]
Phil,
  I don't think you will save much moving wave from inner to outer loop -
maybe 1 node for C384.
If you do move, the coupling is not as good where wave action may be
important; however considering all the approximations involved it isn't
that important.
Moorthi

On Fri, Oct 22, 2021 at 12:21 PM Phil Pegion ***@***.***>
wrote:

> @SMoorthi-emc <https://github.com/SMoorthi-emc> but does the roughness
> length really need to be updated every physics time-step, and do the wave
> need to have winds updated so frequently? Running at the ocean time-step is
> a cost savings of 6x.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ufs-community/ufs-weather-model/discussions/885#discussioncomment-1522250>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ALLVRYRJMARRDPVFMY4Z2ZTUIGFPRANCNFSM5GQVNIKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Dr. Shrinivas Moorthi
Research Meteorologist
Modeling and Data Assimilation Branch
Environmental Modeling Center / National Centers for Environmental
Prediction
5830 University Research Court - (W/NP23), College Park MD 20740 USA
Tel: (301)683-3718

e-mail: ***@***.***
Phone:  (301) 683-3718   Fax: (301) 683-3718

**Comment by junwang-noaa** [comment]
My understanding is that the code will still work as  atm will use the z0 in the fv3 export state until the next time wave updates it, the same for ww3. Maybe we can try to see how much cost this can save.

**Reply by pjpegion** [reply]
@junwang-noaa I'm willing to test, but I may need help with the nems.configure

**Reply by JessicaMeixner-NOAA** [reply]
@pjpegion Maybe something like this:

runSeq::
@1800
MED med_phases_prep_ocn_avg
MED -> OCN :remapMethod=redist
OCN -> WAV
WAV -> OCN :srcMaskValues=1
WAV -> ATM :srcMaskValues=1
ATM -> WAV
ICE -> WAV
OCN
WAV
@300
MED med_phases_prep_atm
MED med_phases_prep_ice
MED -> ATM :remapMethod=redist
MED -> ICE :remapMethod=redist
ATM
ICE
ATM -> MED :remapMethod=redist
MED med_phases_post_atm
ICE -> MED :remapMethod=redist
MED med_phases_post_ice
MED med_phases_prep_ocn_accum
@
OCN -> MED :remapMethod=redist
MED med_phases_post_ocn
MED med_phases_restart_write
@

**Reply by pjpegion** [reply]
@JessicaMeixner-NOAA Thanks!

**Reply by pjpegion** [reply]
@JessicaMeixner-NOAA that did not work,  it seems that the atmosphere model is not happy with having the import state be from an older time-step since it is crashing in the 2nd atmospheric timestep:

ERROR            PET000 fv3_cap.F90:1275 Invalid argument  - NUOPC INCOMPATIBILITY DETECTED: Import Field wave_z0_roughness_length not at current time

I guess since the wave model is not going through the mediator, is cannot handle the different coupling intervals.

**Comment by junwang-noaa** [comment]
I am curious, so ww3 can run with large time step (1800s vs 300s) with the
same high resolution grids?

**Reply by pjpegion** [reply]
looks like the CFL criteria, is about 1.5 hours if the wave grid is also 1/4 degree.

**Reply by junwang-noaa** [reply]
Thanks, Phil. So the winds must be quite small in the wave model.

**Reply by pjpegion** [reply]
I think the winds are just fluxes,  i think it is because the ocean swell travels ~10 m/s

**Reply by junwang-noaa** [reply]
I see, with that speed 1800s time steps should be OK.  Now I am wondering why we use 300s time step :).
